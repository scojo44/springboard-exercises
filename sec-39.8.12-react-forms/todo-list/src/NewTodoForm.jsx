import React, {useState} from "react";

const EMPTY_FORM = {
  task: ''
};

const NewTodoForm = ({addTask}) => {
  function handleSubmit(e) {
    e.preventDefault();
    addTask(formData);
    setFormData(() => {
      return {...EMPTY_FORM};
    });
  }

  function handleChange(e) {
    const {name, value} = e.target;

    setFormData(data => {
      return {
        ...data,
        [name]: value
      };
    })
  }

  const [formData, setFormData] = useState(EMPTY_FORM);

  return (
    <form onSubmit={handleSubmit} className="NewTodoForm">
      <p>
        <label htmlFor="task">Task: </label>
        <input type="text" id="task" name="task" value={formData.task} onChange={handleChange} />
      </p>
      <button type="submit">Add Todo</button>
    </form>
  )
};

export default NewTodoForm;
