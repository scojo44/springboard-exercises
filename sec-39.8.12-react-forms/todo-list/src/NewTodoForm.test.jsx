import { render, fireEvent, queryByLabelText } from '@testing-library/react';
import NewTodoForm from './NewTodoForm';

describe('NewTodoForm Tests', () => {
  it('renders without crashing', () => {
    render(<NewTodoForm addTask={x => true} />);
  });

  it('matches snapshot', () => {
    const {asFragment} = render(<NewTodoForm addTask={x => true} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows the form', () => {
    const {getByLabelText} = render(<NewTodoForm addTask={x => true} />);
    const task = getByLabelText('New Task:');
    expect(task).toBeInTheDocument();
  });

  it('submits the form, and calls the parent function with the formData', () => {
    const addTask = vitest.fn();
    const {getByText, getByLabelText} = render(<NewTodoForm addTask={addTask} />);
    const task = getByLabelText('New Task:');
    const button = getByText('Add Todo');

    fireEvent.change(task, {target: {value: 'Learn React'}});
    fireEvent.click(button);

    expect(addTask).toHaveBeenCalled();
    expect(addTask).toHaveBeenCalledWith({task: 'Learn React'});
  });
});
