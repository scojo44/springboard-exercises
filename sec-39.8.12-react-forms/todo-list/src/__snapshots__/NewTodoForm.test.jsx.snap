// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewTodoForm Tests > matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewTodoForm"
  >
    <label
      for="task"
    >
      New Task: 
    </label>
    <input
      id="task"
      name="task"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Add Todo
    </button>
  </form>
</DocumentFragment>
`;
